File: CSharp/TenMinuteWalk.cs
/* http://www.codewars.com/kata/54da539698b8a2ad76000228/train/haskell
You live in the city of Cartesia where all roads are laid out in a perfect grid. 

You arrived ten minutes too early to an appointment, so you decided to take the opportunity to go for a short walk. 
The city provides its citizens with a Walk Generating App on their phones -- 
everytime you press the button it sends you an array of one-letter strings representing directions to walk 
(eg. ['n', 's', 'w', 'e']). 

You know it takes you one minute to traverse one city block, 
so create a function that will return true if the walk the app gives you will take you exactly ten minutes 
(you don't want to be early or late!) and will, of course, return you to your starting point. 

Return false otherwise.

Note: you will always receive a valid array containing a random assortment of direction letters ('n', 's', 'e', or 'w' only). 
It will never give you an empty array (that's not a walk, that's standing still!).
*/

namespace CSharp.Codewars
{
    using System.Linq;
    using System;
    using System.Collections;
    using System.Collections.Generic;
    using Microsoft.VisualStudio.TestTools.UnitTesting;

    public partial class TenMinuteWalkCs
    {
        public static bool isValidWalk(char[] walk)
        {
            Func<char, int> count = (match) => { return walk.Where(c => c == match).Count(); }; 
            return walk.Length == 10 && count('n') == count('s') && count('e') == count('w');
        }
    }

    [TestClass]
    public class TenMinuteWalkCsTests
    {
        [TestMethod]
        public void TenMinuteWalkCs_ValidWalk_True()
        {
            Assert.AreEqual(true, TenMinuteWalkCs.isValidWalk(new char[] { 'n', 's', 'n', 's', 'n', 's', 'n', 's', 'n', 's' }));
        }

        [TestMethod]
        public void TenMinuteWalkCs_ValidWalk2_True()
        {
            Assert.AreEqual(true, TenMinuteWalkCs.isValidWalk(new char[] { 'n', 's', 'e', 'w', 'n', 's', 'e', 'w', 'n', 's' }));
        }

        [TestMethod]
        public void TenMinuteWalkCs_InValidWalk_false()
        {
            Assert.AreEqual(false, TenMinuteWalkCs.isValidWalk(new char[] { 'n', 's', 'n', 's', 'n', 's', 'n', 's', 'n', 'n' }));
        }

        [TestMethod]
        public void TenMinuteWalkCs_ShortWalk_False()
        {
            Assert.AreEqual(false, TenMinuteWalkCs.isValidWalk(new char[] { 'n', 's' }));
        }

        [TestMethod]
        public void TenMinuteWalkCs_InfiniteWalk_False()
        {
            Assert.AreEqual(false, TenMinuteWalkCs.isValidWalk(new char[] { 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n' }));
        }

        [TestMethod]
        public void TenMinuteWalkCs_TooShortWalk_False()
        {
            Random r = new Random(DateTime.Now.Millisecond);
            char[] nsew = new char[4] { 'n', 's', 'e', 'w' };

            for (int i = 1; i < 10; i++)
            {
                char[] walk = new char[i];
                for (int j = 0; j < i; j++)
                {
                    walk[j] = nsew[r.Next(4)];
                }

                Assert.IsFalse(TenMinuteWalkCs.isValidWalk(walk), new String(walk));
            }
        }

        [TestMethod]
        public void TenMinuteWalkCs_SemiRandomWalk_True()
        {
            for (int n = 0; n < 4; n++)
            {
                var walk = Enumerable.Repeat('s', n)
                    .Concat(Enumerable.Repeat('n', n))
                    .Concat(Enumerable.Repeat('w', 5 - n))
                    .Concat(Enumerable.Repeat('e', 5 - n))
                    .ToArray();

                Assert.IsTrue(TenMinuteWalkCs.isValidWalk(walk), new String(walk));
            }
        }
    }
}

File: FSharp/TenMinuteWalk.fs
namespace FSharp.CodeWars
(* http://www.codewars.com/kata/54da539698b8a2ad76000228/train/haskell
You live in the city of Cartesia where all roads are laid out in a perfect grid. 

You arrived ten minutes too early to an appointment, so you decided to take the opportunity to go for a short walk. 
The city provides its citizens with a Walk Generating App on their phones -- 
everytime you press the button it sends you an array of one-letter strings representing directions to walk 
(eg. ['n', 's', 'w', 'e']). 

You know it takes you one minute to traverse one city block, 
so create a function that will return true if the walk the app gives you will take you exactly ten minutes 
(you don't want to be early or late!) and will, of course, return you to your starting point. 

Return false otherwise.

Note: you will always receive a valid array containing a random assortment of direction letters ('n', 's', 'e', or 'w' only). 
It will never give you an empty array (that's not a walk, that's standing still!).
*)

module TenMinuteWalkFs = 

    let isValidWalk (walk:char[]) =
        let move ch = 
            match ch with
            | 'n' -> (1,0)
            | 's' -> (-1,0)
            | 'e' -> (0,1)
            | 'w' -> (0,-1)
            | _ -> (0,0)

        if walk.Length <> 10 then false else
            Array.map move walk
            |> Array.reduce (fun acc item -> ((fst acc) + (fst item),(snd acc) + (snd item)) )
            |> (fun v -> fst v =0 && snd v = 0)
        



open Microsoft.VisualStudio.TestTools.UnitTesting
open FsUnit.MsTest
open NHamcrest.Core
open TenMinuteWalkFs
open System

[<TestClass>] 
type ``Ten Minute Walk Tests`` ()=

    let r = System.Random(DateTime.Now.Millisecond)

    let buildValidWalk n = 
        List.replicate n 's'
        |> List.append (List.replicate n 'n')
        |> List.append (List.replicate (5-n) 'w')
        |> List.append (List.replicate (5-n) 'e')
        |> List.toArray

    let randomMove n = ['n';'s';'e';'w'].[r.Next(4)]
    let buildRandomWalk n = [|1..n|] |> Array.map randomMove

    [<TestMethod>] member test.
        ``TenMinuteWalkFs Valid walk 1`` ()= 
        isValidWalk [|'n'; 's'; 'n'; 's'; 'n'; 's'; 'n'; 's'; 'n'; 's'|] 
        |> should be True

    [<TestMethod>] member test.
        ``TenMinuteWalkFs Valid walk 2`` ()= 
        isValidWalk [| 'n'; 's'; 'e'; 'w'; 'n'; 's'; 'e'; 'w'; 'n'; 's' |]  
        |> should be True

    [<TestMethod>] member test.
        ``TenMinuteWalkFs Invalid walk`` ()= 
        isValidWalk [| 'n'; 's'; 'n'; 's'; 'n'; 's'; 'n'; 's'; 'n'; 'n' |]  
        |> should be False

    [<TestMethod>] member test.
        ``TenMinuteWalkFs Invalid short walk`` ()= 
        isValidWalk [| 'n'; 's' |]  
        |> should be False
        
    [<TestMethod>] member test.
        ``TenMinuteWalkFs Invalid long walk`` ()= 
        isValidWalk [| 'n'; 's';'n'; 's'; 'n'; 's'; 'n'; 's'; 'n'; 's'; 'n'; 'n';'n'; 's';'n'; 's'; 'n'; 's'; 'n'; 's'; 'n'; 's'; 'n'; 'n';'n'; 's';'n'; 's'; 'n'; 's'; 'n'; 's'; 'n'; 's'; 'n'; 'n';'n'; 's';'n'; 's'; 'n'; 's'; 'n'; 's'; 'n'; 's'; 'n'; 'n'; |]  
        |> should be False

    [<TestMethod>] member test.
        ``TenMinuteWalkFs Semi random walk`` ()= 
            [0..3]
            |> List.map buildValidWalk
            |> List.map (fun walk -> (walk, isValidWalk walk))
            |> List.map (fun r -> printfn "%s %b" (new string(fst r)) (snd r) |> (fun n -> snd r)  )
            |> List.map (should be True)
            |> ignore

    [<TestMethod>] member test.
        ``TenMinuteWalkFs Short random walks`` ()= 
            [1..10]
            |> List.map buildRandomWalk
            |> List.map (fun walk -> (walk, isValidWalk walk))
            |> List.map (fun r -> printfn "%s %b" (new string(fst r)) (snd r) |> (fun n -> snd r)  )
            |> List.map (should be False)
            |> ignore